contracts/marketplace/lib.rs

```diff
@@ -0,0 +1,299 @@
+            self.min_price
+        }
+
+        #[ink(message)]
+        /// Address of the reward token contract this contract will accept as payment.
+        pub fn reward_token(&self) -> AccountId {
+            self.reward_token
+        }
+
+        #[ink(message)]
+        /// Address of the ticket token contract this contract will auction off.
+        pub fn ticket_token(&self) -> AccountId {
+            self.ticket_token
+        }
+
+        #[ink(message)]
+        /// Buy one ticket at the current_price.
+        ///
+        /// The caller should make an approval for at least `price()` reward tokens to make sure the
+        /// call will succeed.
+        pub fn buy(&mut self) -> Result<(), Error> {
+            if self.ticket_balance()? > 0 {
+                let price = self.current_price();
+                let account_id = self.env().caller();
+
+                self.take_payment(account_id, price)?;
+                self.give_ticket(account_id)?;
+
+                self.total_proceeds = self.total_proceeds.saturating_add(price);
+                self.tickets_sold = self.tickets_sold.saturating_add(1);
+                self.current_start_block = self.env().block_number();
+                Self::emit_event(self.env(), Event::Buy(Buy { price, account_id }));
+
+                Ok(())
+            } else {
+                Err(Error::MarketplaceEmpty)
+            }
+        }
+
+        #[ink(message)]
+        /// Re-start the auction from the current block.
+        ///
+        /// Note that this will keep the average estimate from previous auctions.
+        pub fn reset(&mut self) -> Result<(), Error> {
+            Self::ensure_role(self.admin())?;
+
+            self.current_start_block = self.env().block_number();
+            Self::emit_event(self.env(), Event::Reset(Reset {}));
+
+            Ok(())
+        }
+
+        fn current_price(&self) -> Balance {
+            let block = self.env().block_number();
+            let elapsed = block.saturating_sub(self.current_start_block.into());
+            self.average_price()
+                .saturating_mul(self.sale_multiplier)
+                .saturating_sub(self.per_block_reduction().saturating_mul(elapsed.into()))
+                .max(self.min_price)
+        }
+
+        fn per_block_reduction(&self) -> Balance {
+            self.average_price()
+                .saturating_div(self.auction_length.into())
+                .max(1u128)
+        }
+
+        fn take_payment(&self, from: AccountId, amount: Balance) -> Result<(), Error> {
+            build_call::<Environment>()
+                .call_type(Call::new().callee(self.reward_token))
+                .exec_input(
+                    ExecutionInput::new(Selector::new(TRANSFER_FROM_GAME_TOKEN_SELECTOR))
+                        .push_arg(from)
+                        .push_arg(self.this())
+                        .push_arg(amount)
+                        .push_arg(DUMMY_DATA),
```

Comment:
Can you explain why we need this `DUMMY_DATA` arg?

Reply:
there's an argument in `psp22::transfer` that allows for passing some data along with a transfer - for messages, or callbacks

Reply:
But do we need to pass it if it's unused? Can't the `psp22::transfer` make it optional?

Reply:
it's in an external crate :(

Reply:
Thanks.